R/utils.R

```diff
@@ -58,8 +58,16 @@ include_parent_datanames <- function(dataname, join_keys) {
 #' @keywords internal
 teal_data_to_filtered_data <- function(x, datanames = teal.data::datanames(x)) {
   checkmate::assert_class(x, "teal_data")
-  checkmate::assert_character(datanames, min.len = 1L, any.missing = FALSE)
-  checkmate::assert_subset(datanames, teal.data::datanames(x))
+  checkmate::assert_character(datanames, any.missing = FALSE)
+
+  if (length(datanames) == 0) {
+    datanames <- ls(x@env)
+    if (length(datanames) == 0) {
+      stop("cannot assign datanames as `teal_data` environment is empty. Contact app developer.")
+    }
```

Comment:
1. A `stop` on empty `teal_data` should be triggered much earlier, in `init` probably. This is not the place.

Reply:
Re 1: I would argue that it is the right place with the rationale of keeping the logic encapsulated within a given function. If we move this stop condition elsewhere we will spread this specific logic across multiple functions which is a bad code smell for me. The responsibility of `teal_data_to_filtered_data` is to create one object instance on a basis of another and this stop statement is an information that this cannot be done because of some reason - a specific condition for a specific task. For me this looks good.
On the other hand I do understand a fail-fast that you probably thinking of but in this case (of specific condition for a specific action) I would personally value more better code organization.

Reply:
No that `init` does this at the moment:
```
    if (length(teal.data::datanames(data)) == 0) {
      stop("`data` object has no datanames. Specify `datanames(data)` and try again.")
    }
```

I would argue that if an empty object cannot be used to build an app, it should fail immediately, after several processing steps.

The purpose of `teal_data_to_filtered_data` is to take a `teal_data` and turn it into a `filtered_data`. Setting `datanames` on that `teal_data` is not in its purview.

The coming changes to `teal` and `teal.slice` will remove the need for this `stop` anyway but that's another story.

Reply:
```
 if (length(teal.data::datanames(data)) == 0) {
      stop("`data` object has no datanames. Specify `datanames(data)` and try again.")
    }

```
I added an additional condition to check for environment and restored above code. This is to ensure that the app initialization fails if an empty object is passed. Previously, I intended to minimize these checks and keep them in one place. However, I now favor failing the app on initialization for an empty object. committed change here [6866652](https://github.com/insightsengineering/teal/pull/1004/commits/6866652edddbedbfaa3d7fb928ce791aae5e9462).

Reply:
> No that `init` does this at the moment:
> 
> ```
>     if (length(teal.data::datanames(data)) == 0) {
>       stop("`data` object has no datanames. Specify `datanames(data)` and try again.")
>     }
> ```

I have a similar objection to this and probably more in there but that's a separate issue out of scope of this PR.

> I would argue that if an empty object cannot be used to build an app, it should fail immediately, after several processing steps.

You are risking distributing logic across many functions. 
Let's take a following dummy example:
```
foo <- function(a, b) {
  aa <- a_handler(a)
  bb <- b_handler(b)
  bar(aa, bb)
}
```
If I understand you right you are suggesting the following
```
foo <- function(a, b) {
  stopifnot(check1(a))
  stopifnot(check2(a))
  ...
  (same for `b`)
  (body)
}
```
Then you have part of the logic for `a` in `foo` and the other part in `a_handler`. I am thinking to have it into `a_handler` instead. Both are fully functional but differs from code organization perspective. Following single responsibility paradigm - (broadly speaking) I want to change `a_handler` only on `a` change.

PS. I am actually seeing `teal_data_to_filtered_data` as the _main_ handler of `teal_data` class of object.

I think that the good middle ground is something we used to do: to perform _just_ class checks upfront in the top-level function to fail fast but don't put any specific logic outside of handlers.

I am raising this to your attention as recent refactor (in particular: the need of modify ~a lot of places~ almost everything) just exposed previous mistakes. Obviously those were big changes but still it should had been more condensed.

> The purpose of `teal_data_to_filtered_data` is to take a `teal_data` and turn it into a `filtered_data`. Setting `datanames` on that `teal_data` is not in its purview.

Yes but I'm seeing this more of a validating than setting (even though formal set is called).

> The coming changes to `teal` and `teal.slice` will remove the need for this `stop` anyway but that's another story.

Ok then I'm not pushing anymore on this particular case but I really want you to understand the rationale not to repeat previous mistakes.

Reply:
> PS. I am actually seeing `teal_data_to_filtered_data` as the _main_ handler of `teal_data` class of object.

Ah, that is where we fundamentally differ. I think we are unlikely to hammer out a consensus in this form.

> I am raising this to your attention as recent refactor (in particular: the need of modify ~a lot of places~ almost everything) just exposed previous mistakes. Obviously those were big changes but still it should had been more condensed.

We have cut some heads off the this hydra but the battle is not over yet :wink: And since we are changing everything incrementally, we have transition stages. I think this is one.

As for moving a `stop` from one place to another, it does seem out of the scope of this issue. I hate to be a broken record but we have had too much of that recently.